### PR TITLE
Updates from GMAO for GEOS

### PR DESCRIPTION
**Pull request by lizziel**
### Name and Institution (Required)

Name: Lizzie Lundgren
Institution: Harvard University

### Describe the update

Updates include the following:

- Updates impacting the standard offline model:
  - Two new optional diagnostics to track number of negative concentrations after KPP integration
    - State_Diag%KPPNegatives is number of grid cells with negative concentration after KPP integration
    - State_Diag%KPPNegatives0 is number of grid cells with negative concentration after first KPP integration attempt
- GEOS-only updates:
  - Add run-time option to check for negatives after KPP integration (Input_Opt%KppCheckNegatives)
  - Add run-time parameter for KPP tolerance inflation factor for second KPP integration attempt (Input_Opt%KppTolScale) 
  - Turn off 3 heterogeneous reactions in stratosphere by default (previously default was not to do this)
  - Add parameter to define RATS phase
  - Move GEOS diagnostic exports to geos_interface.F90
  - Always prescribe H2O in both the stratosphere and troposphere in GEOS
  - Updates for GEOS H2O tendency and Ox diagnostics
  - Add logic to determine whether to apply H2O tendency to Q
  - Add subroutine to set H2O from specific humidity (GEOS_SetH2O)
  - Use absolute data paths in GEOS-Chem ExtData yaml file used in GEOS
  - Set species friendly to moist by default
  - Add inventory for GEOS 3-hourly methane within PBL
  - Turn off GEOS-Chem convection by default
  - Update Fast-JX input directory
 
### Expected changes

This is a zero diff update for offline GEOS-Chem.

### Reference(s)

None

### Related Github Issue(s)

None

### CODE REVIEW

File: GeosCore/fullchem_mod.F90 [hunk 1]
```diff
@@ -1130,6 +1132,30 @@ SUBROUTINE Do_FullChem( Input_Opt,  State_Chm, State_Diag,                 &
...
+                ENDIF
+             ENDIF
+          ENDIF
+       ENDIF
+       IF ( State_Diag%Archive_KppNegatives0 ) THEN
+          S = 0
+          DO N=1,NSPEC
+             IF ( C(N) < 0.0_dp ) S=S+1
+          ENDDO 
+          State_Diag%KppNegatives0(I,J,L) = REAL(S,KIND=f4)
```
Comment (yantosca) [inline GeosCore/fullchem_mod.F90:1155]: You can get rid of the loop with an array-based function, which will be more efficient than looping:
```F90 
State_Diag%KppNegatives0(I,J,L) = REAL( COUNT( C < 0.0_dp ), KIND=4 )
```

File: GeosCore/fullchem_mod.F90 [hunk 2]
```diff
@@ -1375,6 +1407,13 @@ SUBROUTINE Do_FullChem( Input_Opt,  State_Chm, State_Diag,                 &
           ! Skip if this is not a GEOS-Chem species
           IF ( SpcID <= 0 ) CYCLE
 
+          ! Scan for negatives
+          IF ( State_Diag%Archive_KppNegatives ) THEN
+             IF ( C(N) < 0.0_dp ) THEN 
+                State_Diag%KppNegatives(I,J,L) = State_Diag%KppNegatives(I,J,L) + 1.0_dp
```
Comment (yantosca) [inline GeosCore/fullchem_mod.F90:1413]: The increment should be `1.0_f4` (since the array is REAL(f4), or single precision.  Fortran will cast it anyway, but I think using the proper KIND value prevents the cast.

File: GeosCore/fullchem_mod.F90 [hunk 3]
```diff
@@ -1130,6 +1132,30 @@ SUBROUTINE Do_FullChem( Input_Opt,  State_Chm, State_Diag,                 &
...
+       ! Check for negative concentrations after first integration and 
+       ! eventually mark integration as errorenous so that it will be
+       ! repeated below. cakelle2, 2023/10/26.
+       IF ( Input_Opt%KppCheckNegatives >= 0 ) THEN
+          IF ( ( Input_Opt%KppCheckNegatives==0 .AND. State_Met%InStratMeso(I,J,L) ) .OR. &
+               ( L > (State_Grid%NZ-Input_Opt%KppCheckNegatives) ) ) THEN
+             IF ( ANY(C < 0.0_dp) ) THEN
+                IERR = -999
+                IF ( State_Diag%Archive_KppError ) THEN
+                   State_Diag%KppError(I,J,L) = State_Diag%KppError(I,J,L) + 1.0
```
Comment (yantosca) [inline GeosCore/fullchem_mod.F90:1145]: The increment should be `1.0_f4` since I believe KppError is a single-precision field.

**Review by yantosca — CHANGES_REQUESTED** [review]
These changes look OK to merge.  I've noted some comments for efficiency.

**Comment by christophkeller** [comment]
Good points @yantosca! The suggested changes look good to me, thanks for pointing them out!

File: GeosCore/fullchem_mod.F90 (hunk 3 above)
Comment (lizziel) [reply GeosCore/fullchem_mod.F90:1145]: Done.

**Comment by lizziel** [comment]
@christophkeller, I added another commit on top of this with some minor changes. `State_Diag%KPPnegatives0` is now set outside of the MODEL_GEOS loop so it is available for all configurations of the model. The check for negatives now includes condition IERR >= 0. This is to avoid double-counting when incrementing diagnostic `State_Daig%KppError` which was already incremented if IERR < 0.

You probably noticed I excluded all KPP directory updates. The updates from Eric Fleming have conflicts with JPL updates from Kelvin Bates that have gone into dev/14.3.0. Would it be possible for Eric to make new KPP updates as needed on top of 14.3.0?

**Comment by lizziel** [comment]
@yantosca, thanks for the recommendations! I pushed updates to address all of your comments.

**Comment by christophkeller** [comment]
> @christophkeller, I added another commit on top of this with some minor changes. `State_Diag%KPPnegatives0` is now set outside of the MODEL_GEOS loop so it is available for all configurations of the model. The check for negatives now includes condition IERR >= 0. This is to avoid double-counting when incrementing diagnostic `State_Daig%KppError` which was already incremented if IERR < 0.
> 
> You probably noticed I excluded all KPP directory updates. The updates from Eric Fleming have conflicts with JPL updates from Kelvin Bates that have gone into dev/14.3.0. Would it be possible for Eric to make new KPP updates as needed on top of 14.3.0?

This is fine, I suspected that this might happen. Once released, we can go through 14.3.0 to make sure that all of Eric's updates are being integrated.

**Review by yantosca — APPROVED** [review]
Good to merge!  Thanks @lizziel.

**Comment by lizziel** [comment]
I updated the branch to be off of dev/no-diff-to-benchmark rather than dev/14.3.0. @yantosca, could you take a look again. Should be good to merge once you give the go-ahead.

**Review by yantosca — APPROVED** [review]
Thanks @lizziel.  Good to merge!

**Comment by yantosca** [comment]
@msulprizio @lizziel: Integration tests now running

**Comment by yantosca** [comment]
All GEOS-Chem Classic integration tests passed (except TOMAS):
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

GCClassic #457a610 GEOS-Chem submod update: Merge PR #2095 (Increase time in GCHP int tests)
GEOS-Chem #aad13ea0a Merge PR #2062 (Updates from GMAO for GEOS)
HEMCO     #a5d5169 HEMCO 3.7.2 release

Using 24 OpenMP threads
Number of execution tests: 27

Submitted as SLURM job: 14497727
==============================================================================
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 25
Execution tests failed: 2
Execution tests not yet completed: 0
```
All GEOS-Chem Classic integration tests were zero-diff w/r/t PR #2095 except for:
- TOMAS (failed)
- APM (parallel error)
- RRTMG (numerical noise in RRTMG collection output)

**Comment by yantosca** [comment]
All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

GCHP      #5692f34 GEOS-Chem submod update: Merge PR #2084 (Fix exe name in int tests)
GEOS-Chem #aad13ea0a Merge PR #2062 (Updates from GMAO for GEOS)
HEMCO     #a5d5169 HEMCO 3.7.2 release

Number of execution tests: 5

Submitted as SLURM job: 14497816
==============================================================================
 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Also, all GCHP integration tests were zero-diff w/r/t PR #2095.
